Clean up some EndpointSlice-related code in apiserver endpoint reconciler

danwinship opened the pull request:
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
1. Removes code that deals with the possibility of EndpointSlice being disabled (which can't happen any more).
2. Refactors endpoint reconciler unit tests in preparation for flipping the reconciler from "an Endpoints reconciler that also outputs EndpointSlices" to "an EndpointSlice reconciler that also outputs Endpoints", which will be needed for dual-stack endpoint reconciling.
3. ~Removes code to deal with old objects that are so old we couldn't possibly actually encounter them any more.~ (no longer in latest version)
4. ~Removes code to deal with Endpoint/EndpointSlice translation of fields that we never actually ended up using in the apiserver Endpoints.~ (no longer in latest version)

#### Which issue(s) this PR fixes:
None; I'm doing some cleanup/refactoring of the endpoint reconciling code in preparation for dual-stack apiserver support (KEP-2438).

#### Special notes for your reviewer:
(The first commit also appears in #107872 but either PR can merge without the other.)

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/2438-dual-stack-apiserver
```

/sig api-machinery

danwinship committed:
Add more Endpoints vs EndpointsSlices reconciler tests

danwinship commented:
/cc @robscott

aojea commented:
> 3\. Removes code to deal with Endpoint/EndpointSlice translation of fields that we never actually ended up using in the apiserver Endpoints.

[I think we should keep that code that makes the implementation more correct ](https://github.com/kubernetes/kubernetes/pull/107878/commits/08cb23790fb22d5f173ae1a816aacd00b1078bb3) and avoid simplifying based on our implementation details, if we decide to use those fields later do we have to add again that code ?

Another important detail is that the Conformance doesn't specify who has to set the apiserver endpoints, there are Kubernetes implementations that create the apiserver endpoints externally, not using the reconcilers, a corner case but possible, is that one of those distros upgrade and start using the apiserver reconcilers ...

danwinship commented:
> if we decide to use those fields later do we have to add again that code ?

I'm not sure we have any reason to believe that the code that is currently there is even correct.

> there are Kubernetes implementations that create the apiserver endpoints externally

But they would have to use `--endpoint-reconciler-type=none` wouldn't they? In which case this code doesn't run at all.

aojea commented:
> I'm not sure we have any reason to believe that the code that is currently there is even correct.

then is a bug :)

> But they would have to use `--endpoint-reconciler-type=none` wouldn't they? In which case this code doesn't run at all.

I didn't check, but is it possible if 1..X reconciler off ->  1.x+1 reconciler on?

fedebongio commented:
/assign @aojea 
Antonio looks like you are already on this one? Thank you!
/triage accepted

aojea commented on pkg/controlplane/reconcilers/endpointsadapter.go:
```diff
@@ -93,19 +93,8 @@ func (adapter *EndpointsAdapter) EnsureEndpointSliceFromEndpoints(namespace stri
-
 	if apiequality.Semantic.DeepEqual(currentEndpointSlice.Endpoints, endpointSlice.Endpoints) &&
-		apiequality.Semantic.DeepEqual(currentEndpointSlice.Ports, endpointSlice.Ports) &&
-		apiequality.Semantic.DeepEqual(currentEndpointSlice.Labels, endpointSlice.Labels) {
```
we want to check the skip-mirror label is set to reconcile and avoid people remove it?
 related to https://github.com/kubernetes/kubernetes/issues/104252 ?

robscott replied:
Yeah I think all labels should be consistent here, is there any reason we'd want to remove this check?

danwinship replied:
> we want to check the skip-mirror label is set

(The `skip-mirror` label is on the Endpoints, not the EndpointSlice; this check is about `kubernetes.io/service-name`.)

> to reconcile and avoid people remove it? related to #104252 ?

The existing apiserver code does not try to repair intentional sabotage. It only tries to fix objects that are in a previously-valid state.

> I think all labels should be consistent here, is there any reason we'd want to remove this check?

The apiserver always creates consistent labels already, so the only way they could be inconsistent would be if the administrator was manually trying to break it.

(We _could_ keep the Labels check here, but if so we should probably check Annotations, AddressType, and maybe a few other things too.)

aojea replied:
> (The skip-mirror label is on the Endpoints, not the EndpointSlice; this check is about kubernetes.io/service-name.)

duh :upside_down_face: , my bad

> > to reconcile and avoid people remove it? related to #104252 ?
> 
> The existing apiserver code does not try to repair intentional sabotage. It only tries to fix objects that are in a previously-valid state.

yeah, but it seems that the reason to not do that, was the multiport feature that you are going to remove ... should we solve this now?

danwinship replied:
I don't know that he was talking about the _existing_ multi-port feature, but more about the problem of apiservers having different versions and/or configs in general. The apiservers have a protocol for reaching agreement on the set of endpoint IPs, but it is assumed/required that they already agree on what the value of every other Service/Endpoints field should be. If they don't, there will be edit wars if each apiserver regularly attempts to assert its own values.

aojea replied:
ok, let me write it down because I recognize I wasn't doing my homework correctly and I was confused and maybe we are talking about different reconcilers:

- The apiserver creates a service "kubernetes.default" and it must have endpoints so cluster members can reach it, it uses this setting to configure the IP address on the endpoint
```
	// PublicAddress is the IP address where members of the cluster (kubelet,
	// kube-proxy, services, etc.) can reach the GenericAPIServer.
	// If nil or 0.0.0.0, the host's default interface will be used.
	PublicAddress net.IP
```

- The apiserver has 3 types or reconcilers for these endpoints
```
	// MasterCountReconcilerType will select the original reconciler
	MasterCountReconcilerType Type = "master-count"
	// LeaseEndpointReconcilerType will select a storage based reconciler
	LeaseEndpointReconcilerType = "lease"
	// NoneEndpointReconcilerType will turn off the endpoint reconciler
	NoneEndpointReconcilerType = "none"
```

*None*: endpoints are generated externally, we don't care about this
*Master*: legacy, it depends on knowing beforehand the number of apiserver running
```
	// Number of masters running; all masters must be started with the
	// same value for this field. (Numbers > 1 currently untested.)
	MasterCount int
```
The reconciliation is based on the Endpoint object, it checks its own IP is in the object and the number of addresses is lower or equal to MasterCount


*Lease*: the default, each apiserver writes a lease directly in etcd and renews it periodically.
```
/registry/masterleases/172.18.0.4
/registry/masterleases/172.18.0.5
/registry/masterleases/172.18.0.6
```

The reconciliation is based on the IPs listed in the leases match the ones in the endpoint, and the lease TTL has renewed on time (there is an option to set the TTL expiration time)

I think we should start deprecating "Master" reconciler as it is easy to get conflicts, just if you don't set the MasterCount setting correctly.

If we consider only the Lease reconciler, I think that we can solve the problem of external actors modifying the apiserver endpoints.

danwinship replied:
> I think we should start deprecating "Master" reconciler as it is easy to get conflicts, just if you don't set the MasterCount setting correctly.

Yeah, I already suggested in the dual stack apiserver KEP that we'd only support dual-stack apiservers with the `lease` reconciler.

> If we consider only the Lease reconciler, I think that we can solve the problem of external actors modifying the apiserver endpoints.

If we _only_ care about bad external actors we can solve it now, with either reconciler; each apiserver can just resync the ports value whenever it sees that the current value does not match what it thinks it should be.

The reason we don't do that (I assume) is because it fails when you're trying to change the config (or when a new version of kubernetes adds a new feature that changes the desired output; eg, if we want to add QUIC / HTTP3 support).

But yes, we can fix that case easily with the `lease` reconciler but not with the `master-count` reconciler; we just make the reconciler write out the ports to the lease objects as well as the IPs, and then adopt the rule that if all active masters agree on the value of the ports, then it should sync that value, but whenever there is disagreement between masters, it should leave the ports unchanged. (This matches the behavior we agreed would work best for dual-stack upgrade/rollback; the apiserver is dual-stack only when every apiserver writes out a dual-stack lease.)

(And we can't easily implement that same behavior with the `master-count` reconciler, because that reconciler doesn't have a separate set of objects where each apiserver can indicate its own theory of what the ports should be; the object that they use to communicate with each other is the `default/kubernetes` Endpoints object itself.)

But anyway, if we want to do this, this is a separate thing which is not incompatible with this PR. (And, as kind of implied by one of my other comments, in order to make dual-stack work, we'll have to flip the reconciler code to treat the EndpointSlices as the "real" output and generate the Endpoints from the EndpointSlices, rather than vice versa, so this is all going to get rewritten anyway.)

danwinship replied:
> But anyway, if we want to do this, this is a separate thing which is not incompatible with this PR.

though also, I could just remove the "Remove some dead EndpointsAdapter code" commit from this PR, and have this be _just_ about making EndpointSlices required rather than optional

aojea replied:
that will be faster, that commit is the one creating doubts, for me, main concern is removing code that we can't easily predict the fallout ... this is a very sensitive change

aojea commented:
@danwinship my main concern was about the multiport feature, but since this is going away https://github.com/kubernetes/kubernetes/pull/107872 it LGTM, just want comment

robscott commented on pkg/controlplane/controller_test.go:
```diff
@@ -34,567 +33,6 @@ import (
 	netutils "k8s.io/utils/net"
 )
 
-func TestReconcileEndpoints(t *testing.T) {
```
Just to make sure I'm understanding this PR correctly, these tests were moved, not deleted, right?

danwinship replied:
correct

robscott commented on pkg/controlplane/reconcilers/lease_test.go:
```diff
@@ -355,43 +355,21 @@ func TestLeaseEndpointReconciler(t *testing.T) {
 				}},
 			},
 		},
-		{
-			testName:      "existing endpoints without skip mirror label",
-			serviceName:   "foo",
-			ip:            "1.2.3.4",
-			endpointPorts: []corev1.EndpointPort{{Name: "foo", Port: 8080, Protocol: "TCP"}},
-			endpoints: &corev1.EndpointsList{
-				Items: []corev1.Endpoints{{
-					ObjectMeta: om("foo", false),
-					Subsets: []corev1.EndpointSubset{{
-						Addresses: []corev1.EndpointAddress{{IP: "1.2.3.4"}},
-						Ports:     []corev1.EndpointPort{{Name: "foo", Port: 8080, Protocol: "TCP"}},
-					}},
-				}},
-			},
```
Removing this because the idea is that relevant Endpoints would have to already have this label at this point?

danwinship replied:
right

robscott commented on pkg/controlplane/reconcilers/endpointsadapter.go:
```diff
@@ -139,42 +123,37 @@ func endpointSliceFromEndpoints(endpoints *corev1.Endpoints) *discovery.Endpoint
-		if allAddressesIPv6(append(subset.Addresses, subset.NotReadyAddresses...)) {
+		if allAddressesIPv6(subset.Addresses) {
 			endpointSlice.AddressType = discovery.AddressTypeIPv6
 		}
 
-		endpointSlice.Endpoints = append(endpointSlice.Endpoints, getEndpointsFromAddresses(subset.Addresses, endpointSlice.AddressType, true)...)
-		endpointSlice.Endpoints = append(endpointSlice.Endpoints, getEndpointsFromAddresses(subset.NotReadyAddresses, endpointSlice.AddressType, false)...)
+		endpointSlice.Endpoints = append(endpointSlice.Endpoints, getEndpointsFromAddresses(subset.Addresses, endpointSlice.AddressType)...)
```
This seems like we're unnecessarily losing data? Why would we publish unready endpoints with the Endpoints API but not EndpointSlice?

danwinship replied:
We _don't_ publish unready endpoints with the Endpoints API. That's why I'm removing this code; it never does anything, because `endpoints.NotReadyAddresses` is always empty.

(We _can't_ publish unready endpoints, because the `reconcilers.EndpointReconciler` interface doesn't provide any way for the apiserver to indicate to it that it's running but not ready. And if we wanted to extend it to do this in the future, we would want to be able to distinguish terminating from non-terminating as well, but there's no way to represent that in an Endpoints, so we would have to also flip the reconciler implementations around so that the EndpointSlice was their "primary" output, and the Endpoints was generated from the EndpointSlice, rather than the other way around like it is now. So even then, we wouldn't want _this_ code; we'd need _the opposite_ code, to generate `NotReadyEndpoints` in the Endpoints based on the `Conditions` in the EndpointSlice.)

robscott commented on pkg/controlplane/reconcilers/endpointsadapter.go:
```diff
@@ -139,42 +123,37 @@ func endpointSliceFromEndpoints(endpoints *corev1.Endpoints) *discovery.Endpoint
-
-	if address.NodeName != nil {
-		ep.NodeName = address.NodeName
 	}
```
Is this being removed because it's never populated?

danwinship replied:
correct

robscott reviewed (commented):
Thanks for the work on this @danwinship!

danwinship commented:
Updated. The only non-unit-test changes now are:
- some code got moved
- EndpointsAdapter now assumes EndpointSlices are enabled rather than optionally ignoring EndpointSlices
- endpointSliceFromEndpoints now explicitly sets ObjectMeta.Namespace, which is needed to make some comparisons come out right in the unit tests (since the fake client doesn't fill it in)

I also updated the unit tests some more to more explicitly test both Endpoints and EndpointSlices (which adds a bunch of churn now but will reduce churn later when we flip the reconcilers around to work primarily with EndpointSlice rather than Endpoints)

danwinship commented:
/assign @deads2k 
(@aojea want to re-review and lgtm it? Everything you didn't like before should be gone now)

aojea commented on pkg/controlplane/reconcilers/instancecount_test.go:
```diff
@@ -564,21 +344,18 @@ func TestMasterCountEndpointReconciler(t *testing.T) {
-		}},
-	}
-	fakeClient := fake.NewSimpleClientset()
-	if endpoints != nil {
```
heh

aojea commented on pkg/controlplane/reconcilers/endpointsadapter_test.go:
```diff
@@ -37,59 +38,52 @@ func TestEndpointsAdapterGet(t *testing.T) {
 		expectedError         error
 		expectedEndpoints     *corev1.Endpoints
-		endpoints             []*corev1.Endpoints
+		initialState          []runtime.Object
```
ok, this is runtime.Object because you mix endpoints and endpointslices on the client initialization

aojea commented on pkg/controlplane/reconcilers/helpers_test.go:
```diff
@@ -41,3 +48,51 @@ func makeEndpoints(name string, ips []string, ports []corev1.EndpointPort) *core
 }
+
+func verifyCreatesAndUpdates(fakeClient *fake.Clientset, expectedCreates, expectedUpdates []runtime.Object) error {
+	errors := []error{}
```
just a comment, maybe you already know, golang has this `t.Helper()` method for these kind of functions

aojea commented:
/lgtm

cleans up and simplifies the code on tests

you need David for approval

danwinship committed:
Simplify endpoint creation in reconciler unit tests

Also make the expectCreate / expectUpdate fields into arrays while
we're rewriting their values anyway, to avoid additional churn in the
next commit.

danwinship committed:
Simplify reconciler unit test setup

Pass initial state objects to fake.NewSimpleClientSet() rather than
calling Create() by hand.

(This will make it easier to have an initial state that is a mix of
Endpoints and EndpointSlices later on.)

danwinship committed:
Use t.Run() consistently in reconciler unit tests

danwinship committed:
Improve reconciler test result checking

Consistently verify creates/updates based on the fake client's action
tracking, not based on the return values of the reconciler functions.
(This will also let us check that both Endpoints and EndpointSlices
were created/updated correctly.)

danwinship committed:
Use EndpointSlices in all reconciler unit tests

EndpointSlice is always enabled now, so make it non-optional in the
EndpointsAdapter, make all of the test cases pass an EndpointSlice
client, and remove the "EndpointSlices disabled"-specific tests.

By changing makeEndpointsArray() to return both an Endpoints and an
EndpointsSlice, the "initialObjects" and "expectCreate"/"expectUpdate"
fields of (almost) all of the existing unit tests are automatically
switched to be EndpointSlice-aware; instead of having an initial state
with just Endpoints (or nothing), and testing that just the Endpoints
is created/updated correctly, they now have an initial state with both
Endpoints and EndpointSlice (or nothing), and test that both objects
are created/updated correctly.

The handful of existing test cases that used inconsistent Endpoints
and EndpointSlice state have been modified to create the objects
separately.

danwinship committed:
Remove unused endpointSlicesEnabled fields

deads2k commented on pkg/controlplane/reconcilers/lease_test.go:
```diff
@@ -81,318 +79,187 @@ func (f *fakeLeases) Destroy() {
 			endpointPorts: []corev1.EndpointPort{{Name: "foo", Port: 8080, Protocol: "TCP"}},
-			endpoints: &corev1.EndpointsList{
-				Items: []corev1.Endpoints{{
-					ObjectMeta: om("foo", false),
```
found it!

deads2k commented:
Now that's a lot of tests. The appear equivalent and improved to me.  the code change is minimal.

/lgtm
/approve
